core/src/main/scala/org/http4s/Status.scala

```diff
@@ -76,102 +77,123 @@ object Status {
     val ServerError = Status.Informational
   }
 
-  private def mkStatus(code: Int, reason: String = ""): ParseResult[Status] =
-    if (code >= 100 && code <= 599)
-      ParseResult.success(Status(code, reason, isEntityAllowed = true))
-    else ParseResult.fail("Invalid status", s"Code $code must be between 100 and 599, inclusive")
-
-  private def lookup(code: Int) =
-    if (code < 100 || code > 599) None else Option(registry.get(code))
+  val MinCode = 100
+  val MaxCode = 599
 
-  def fromInt(code: Int): ParseResult[Status] = lookup(code).getOrElse(mkStatus(code))
+  def fromInt(code: Int): ParseResult[Status] = lookup(code) match {
+    case None => parseAsStatus(code)
+    case Some(status) => Right(status)
+  }
 
   def fromIntAndReason(code: Int, reason: String): ParseResult[Status] =
-    lookup(code).filter(_.right.get.reason == reason).getOrElse(mkStatus(code, reason))
+    lookup(code, reason) match {
+      case None => parseAsStatus(code, reason)
+      case Some(status) => Right(status)
+    }
 
-  // scalastyle:off magic.number
-  private val registry =
-    new AtomicReferenceArray[Right[Nothing, Status]](600)
-  // scalastyle:on magic.number
+  def registered: Iterable[Status] = all
+
+  private def parseAsStatus(code: Int, reason: String = ""): ParseResult[Status] =
+    if (isStandard(code)) ParseResult.success(Status(code, reason))
+    else
+      ParseResult.fail("Invalid status", s"$code is not a valid response code.")
+
+  private object Registry {
+    private val registry = Array.fill[Option[Status]](MaxCode + 1)(None)
```

Comment:
Could we microoptimize this by filling it with `Either[SomeDummyValue, Status]`?  Then we could return the same reference without unwrapping an `Option` and recreating a `Right`.